Conner Delvaux

Spotify API Data Report

I hypothesize that podcasts that have a longer average duration will be more generally popular in the United States. While there is no way for me to get popularity ratings using the API, I can look at the Spotify charts and use the ID of whatever podcast rating I want. I plan to use the top 1,2 and 3 podcasts and compare the average duration to the number 14, 15, and 16 podcasts.

In [287]:
import urllib
import requests
import pandas as pd
import json
import base64

In [288]:
def get_session_token(SessionID, SessionKey):
    url = 'https://accounts.spotify.com/api/token'
    data = {'grant_type':'client_credentials'}
    encoded_key = base64.b64encode(str(SessionID + ":" + SessionKey).\
                                   encode("ascii"))
    header = {'Authorization': 'Basic {}'.format(encoded_key.decode("ascii"))}
    response = requests.post(url,
                            data = data,
                            headers = header)
    print(response.status_code)
    return response.json()['access_token']

The source of the data is coming from the Spotify API. I not only gives data on podcasts but also music, artists, and albums.

The end point I am going to utilize is the shows/episode endpoint. With it I can gain accces to data such as podcast name, description, sample link, and duration. For this analysis I will only be using name and duration.

In [289]:
keys = pd.read_csv(r"C:\Users\conne\Downloads\Spotify-keys.txt")

access_token = get_session_token(keys['Client_ID'].iloc[0], keys ['Client_secret'].iloc[0])

200


In [290]:
def api_call(endpoint, access_token):
    headers = {'Authorization': f'Bearer {access_token}'}
    response = requests.get(endpoint, headers=headers)
    return response.json()

The top 3 podcasts in the US as of 10/10/24 are:\
Call Her Daddy, hosted by Alex Cooper \
The Joe Rogan Experience, hosted by Joe Rogan\
The Shawn Ryan show, hosted by Shawn Ryan


In [291]:
podcast_id_1 = '7bnjJ7Va1nM07Um4Od55dW'
podcast_id_2 = '4rOoJ6Egrf8K2IrywzwOMk'
podcast_id_3 = '5eodRZd3qR9VT1ip1wI7xQ'

In [292]:
episodes_endpoint_1 = 'https://api.spotify.com/v1/shows/{podcast_id_1}/episodes'
episodes_endpoint_2 = 'https://api.spotify.com/v1/shows/{podcast_id_2}/episodes'
episodes_endpoint_3 = 'https://api.spotify.com/v1/shows/{podcast_id_3}/episodes'

In [293]:
episodes_df_1 = pd.DataFrame(podcast_response_1['items'])
episodes_df_2 = pd.DataFrame(podcast_response_2['items'])
episodes_df_3 = pd.DataFrame(podcast_response_3['items'])

In [294]:
all_episodes_df = pd.concat([episodes_df_1, episodes_df_2, episodes_df_3 ], axis=0)

In [295]:
select_columns_df = all_episodes_df[['name', 'duration_ms']]
select_columns_df


name  duration_ms
0                        Vice President Kamala Harris      2667434
1       Remi Bader: Don’t Call Me Body Positive (FBF)      3255146
2   Heather McMahan: Blow jobs, hall passes, & fra...      4426837
3                              Is He the One? [VIDEO]      2024661
4                    Esther Perel: Cheating 101 (FBF)      3016542
5   Getting Older, Hotter, & Wiser (ft. Heidi Klum...      2801216
6             My Partner is Giving Me the Ick [VIDEO]      2163882
7        Amanda Knox: “I didn’t fucking do it.” (FBF)      5396532
8                     He Cheated, Do I Leave? [VIDEO]      3933651
9   Stassi Schroeder: A Proud Basic Bitch (FBF) [V...      3818133
10    Chris Olsen: My Revenge Porn Experience [VIDEO]      5286997
11                          Anna Delvey (FBF) [VIDEO]      3348906
12  Katy Perry: Narcissists, Blow Jobs & Red Flags...      3543978
13  Lil Dicky: Learning to Love My Dick (FBF) [VIDEO]      3294400
14  The Hamptons Confessional (ft. Noah Centineo, ...      2282708
15  Cole Sprouse: Exploited for Money and Fame (FB...      3362346
16                   I Got Catfished in Paris [VIDEO]      4015765
17                                  Miley Cyrus (FBF)      3473162
18  Owen Thiele: Threesomes, Facetune & Ozempic [V...      5027348
19  Shay Mitchell: Let Him Go, You Deserve Better ...      3654954
0                       #2211 - Michael Shellenberger     11729150
1             #2210 - Calley  Means & Casey Means, MD      8442260
2              JRE MMA Show #164 with Israel Adesanya      9313470
3                                #2209 - Paul Rosolie     12573801
4                              #2208 - Brigham Buhler     11214228
5                                  #2207 - Shawn Ryan      9611860
6                        #2206 - Chamath Palihapitiya      9964798
7                            #2205 - Legion of Skanks     11169086
8                                  #2204 - Matt Walsh      9442878
9                 #2203 - Eric Goode & Jeremy McBride      8768723
10                                 #2202 - Tom Segura      9025214
11                JRE MMA Show #163 - Protect Ya Neck     11194110
12                             #2201 - Robert Epstein      9417918
13                                  #2200 - Kat Timpf     10211177
14                               #2199 - Chris Harris     10645545
15                             #2198 - Bret Weinstein     11726526
16              JRE MMA Show #162 with Belal Muhammad      9755177
17                                 #2197 - Mike Baker      9872105
18                           #2196 - Greg Fitzsimmons     10840020
19                            #2195 - Andrew Huberman     11250345
0   #138 Aerial Recovery - The Hurricane Helene Di...      4722886
1   #137 Avi Loeb - Detecting Remnants of Alien Te...     13215817
2   #136 Cliff Sims - White House Myths, Doomsday ...     10600019
3   #135 Scott Mann - The Real Cost of America's F...      9044688
4   #134 John Gentry - Unpacking the Information W...     10184045
5   #133 John "Shrek" McPhee - The Sheriff of Baghdad     12516310
6   #132 Mike Benz - Inside the Censorship Industr...      9957825
7   #131 Tulsi Gabbard - The Dark Behaviors of the...      7533374
8   #130 JD Vance - Why Have a Government if it's ...      3517622
9      #129 Megyn Kelly - The End of Mainstream Media     13980760
10  #128 Kash Patel - FBI Spying on Americans, Rus...      6462955
11  #127 President Donald J. Trump - Make America ...      4299128
12  #126 Joe Kent - Gold Star Husband and Ex-Speci...     14040737
13  #125 Gina Carano - Disney Crumbles After Manda...     13563768
14  #124 Captain Brad Geary - Inside the Failed In...     21244786
15  #123 Erik Prince - Breakdown of the Donald Tru...      6198491
16  #122 Edwin C. May - Psychics in Space, Dream T...     13085257
17  #121 Cmdr. Ahmad Massoud - The Assassination t...      7312587
18   #120 Annie Jacobsen - Nuclear Armageddon in 2024     13845420
19  #119 Sheriff Mark Lamb - 

I was having a lot of trouble trying to display all the data in this data frame due to using select_columns_df.head() instead of just select_columns_df



In [296]:
average_duration_ms = select_columns_df['duration_ms'].mean()
average_duration_minutes = average_duration_ms // 60000

Learned how to use .mean here: https://saturncloud.io/blog/what-is-pandas-mean-for-a-certain-column/#:~:text=In%20Pandas%2C%20the%20mean(),mean()

then I just divided it by 60000, since there are 60000 ms in a minute. Im not sure how to directly divide the numbers in the data frame.

In [297]:
print("The top 3 Spotify podcasts in the US are an averge of", average_duration_minutes, "minutes long")

The top 3 Spotify podcasts in the US are an averge of 134.0 minutes long


Now i am going to do the same thing with ranked 14th, 15th, and 16th Spotify podcasts and compare the two averages. If my original hypothesis is correct, the average duration will be less than the top 3 average duration.  

The top 14th, 15th, and 16th podcasts in the US as of 10/10/24 are:

This Past Weekend w/Theo Von, hosted by Theo Von (My favorite podcast to listen to on my commute to school)\
The Diary Of CEO with Steven Bartlett, hosted by Steven Bartlett\
The Mel Robbins Podcast, hosted by Mel Robbins


In [298]:
podcast_id_14 = '6PwE1CIZsgfrhX6Bw96PUN'
podcast_id_15 = '7iQXmUT7XGuZSzAMjoNWlX'
podcast_id_16 = '7vz4RYsD5MulTCrcH478t1'

In [299]:
episodes_endpoint_14 = 'https://api.spotify.com/v1/shows/{podcast_id_14}/episodes'
episodes_endpoint_15 = 'https://api.spotify.com/v1/shows/{podcast_id_15}/episodes'
episodes_endpoint_16 = 'https://api.spotify.com/v1/shows/{podcast_id_16}/episodes'

In [300]:
episodes_df_14 = pd.DataFrame(podcast_response_14['items'])
episodes_df_15 = pd.DataFrame(podcast_response_15['items'])
episodes_df_16 = pd.DataFrame(podcast_response_16['items'])

In [301]:
groupB_all_episodes_df = pd.concat([episodes_df_14, episodes_df_15, episodes_df_16 ], axis=0)

Below i am again maing it so only the name and 

In [302]:
groupB_select_columns_df = groupB_all_episodes_df[['name', 'duration_ms']]
groupB_select_columns_df

name  duration_ms
0                                E537 Miranda Lambert      4979435
1               E536 Teamsters President Sean O'Brien      7633293
2                                      E535 Joe Gatto      6983575
3                                  E534 Brooks & Dunn      5865299
4                                     E533 Mark Cuban      7246080
5                                 E532 Dr. Tara Swart      6467657
6                                    E531 Kevin Smith      9055712
7                                      E530 Bobby Lee      5988205
8                                    E529 Post Malone      7609182
9                                 E528 Dermot Kennedy      5638791
10                                       E527 Lobster      5031366
11                                  E526 Donald Trump      3453178
12                                  E525 Jessie Murph      5813655
13                           E524 Sen. Bernie Sanders      3853635
14                             E523 Georges St-Pierre      6413871
15                                     E522 Tommy Lee      7272489
16                                E521 Bassem Youssef      6879712
17                             E520 Rabbi David Wolpe      5757466
18                                   E519 Pauly Shore      6580950
19                                    E518 A Mall Cop      5958373
0   (EXCLUSIVE) Boris Johnson: "They Were Looking ...      6789120
1   The Groundbreaking Cancer Expert: (New Researc...      5828885
2   Moment 182: How To Make People Respect You In ...       597890
3   The No.1 Eye Doctor: They’re Lying To You Abou...      6076522
4   The Brain Rot Dr: How We Produced Millions Of ...      9433962
5   Moment 181: 5 Things Your Body Is Probably Mis...       735582
6   The Anti-Obesity Doctor: If You Don't Exercise...      6785600
7   The Anti-Woke Expert: “We Are Witnessing The F...      7034645
8   Moment 180: Neuroscientist's Top 5 WORST Habit...       534360
9   Glucose Goddess: These 10 Hacks To Strip Fat W...      5923818
10  The Gottman Doctors: Affairs Can Actually Save...      8136960
11  Moment 179: How To Stop Stress From Killing Yo...      1065064
12  The Muscle Building Expert: “Creatine Loading ...      7038933
13  The Cancelled Professor: Husbands Are More Dan...     11027093
14  Moment 178: They're Lying To About Low Fat Foo...       547317
15  Yuval Noah Harari: This Election Will Tear The...      6856042
16  David Epstein: 10,000 Hours Is A Lie! The Morn...      7843498
17  Moment 177: You're NOT Lazy! This Is The REAL ...       689763
18  Andrew Huberman: You Must Control Your Dopamin...     14516736
19  The Dating Doctor: "Start Dating Like It's You...      8442496
0   Don’t Argue or Fight With a Difficult Person –...      3179467
1   How to Learn From Your Mistakes: Let Go of Reg...      3426690
2   Change Your Brain: #1 Neuroscientist’s Exercis...      3964160
3   Simple Life Hacks That Will Change Your Future...      2917851
4   Communicate with Confidence: The Blueprint for...      3746951
5   9 Small Things That Will Make a Surprisingly B...      2742099
6   How to Become a Better Human: Lessons on Kindn...      3530736
7   Living With Purpose: Timeless Wisdom for a Mor...      3980251
8       How to Make a Decision You Won’t Regret Later      2577188
9   The Ultimate Advice for Your Next Chapter (Aft...      4422243
10  The Secret to a Happy Life: What the Longest S...      3219461
11    How to Get Ahead of 99% of People (In 4 Months)      2831438
12  How to Get Motivated Even When You Don’t Feel ...      4012878
13                  8 Things I Wish I Knew in College      2844107
14  7 Lessons To Learn Once That Will Improve Your...      2911764
15  The Science of Women’s Health: Ob/Gyn Reveals ...      3944986
16  Pull Yourself Together: The Best Expert Advice...      3938063
17  8 Realistic Healthy Habits That Make a Huge Di...      2929737
18  How to Figure Out What You Really Want: Use Th...      2763964
19  The “Bronze Medal Mindset

In [303]:
groupB_average_duration_ms = groupB_select_columns_df['duration_ms'].mean()
groupB_average_duration_minutes = groupB_average_duration_ms // 60000

print("The average duration for the 14th, 15th, and 16th ranked Spotify podcast is", groupB_average_duration_minutes, "minutes long")

The average duration for the 14th, 15th, and 16th ranked Spotify podcast is 85.0 minutes long


What the Data showed from both data frames is that the top 3 Spotify podcasts in the US have an average duration of 134 minutes. While the 14th, 15th, and 16th-ranked podcasts averaged a duration of 85 minutes. This analysis supports my original hypothesis that longer average duration podcasts will be more generally popular. Although I don't think this confirms my hypothesis, I believe I would need to use much more data to confirm it. 

My biggest struggle and limitation is not having access to popularity ratings from the Spotify API. I believe this would have made it much easier for me and would have made the Dataframes look better. 